ls1intum/Artemis
# `Communication`: Allow instructors to activate communication and messaging features separately

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
Artemis supports two information-sharing features. `Communication`(Previously called question and answers) and `Messaging`. Currently, the instructor can only deactivate both features at once. This PR allows instructors to activate/deactivate the two features separately.
Closes #6136.

### Description
After discussing with @krusche we decided to change the existing `postsEnabled` flag in the course entity to an enum with four possible values: 

```
    @Enumerated(EnumType.ORDINAL)
    @Column(name = "info_sharing_config")
    @JsonView(QuizView.Before.class)
    private CourseInformationSharingConfiguration courseInformationSharingConfiguration;
```


```
package de.tum.in.www1.artemis.domain.enumeration;

/**
 * NOTE: Ordinal mapping in property of Course.java do NOT change the order of the existing values.
 */
public enum CourseInformationSharingConfiguration {
    /**
     * Both Communication and Messaging are disabled VALUE = 0
     */
    DISABLED,

    /**
     * Both Communication and Messaging are enabled VALUE = 1
     */
    COMMUNICATION_AND_MESSAGING,
    /**
     * Only Communication is enabled VALUE = 2
     */
    COMMUNICATION_ONLY,
    /**
     * Only Messaging is enabled VALUE = 3
     */
    MESSAGING_ONLY,
}

```
For performance reasons, we save this using ordinal mapping on a tiny int column in the db. 

### Steps for Testing
The easiest way to check this from an end-user perspective is to activate/deactivate the respective feature by updating the course and then checking if the navigation tabs 'Communication' or 'Messaging' are hidden in the course overview page.
<img width="1464" alt="image" src="https://user-images.githubusercontent.com/29718932/219624623-34a53984-2fc6-4780-8a82-994c39bee4d5.png">

### Review Progress

#### Code Review
#### Manual Tests
### Test Coverage
Adapted the existing tests.

| Component                      | Statement Coverage | Branch Coverage |
|--------------------------------|--------------------|-----------------|
| ChannelResource                | 100%               | 75%             |
| ConversationManagementResource | 100%               | 100%            |
| ConversationResource           | 100%               | 71%             |
| GroupChatResource              | 100%               | 74%             |
| OneToOneChatResource           | 100%               | 81%             |


### Screenshots

Updated course form:
<img width="1164" alt="image" src="https://user-images.githubusercontent.com/29718932/219623435-2b65e8a0-95c7-4dc2-8722-b159a8a4046a.png">
Tooltip:
<img width="1145" alt="image" src="https://user-images.githubusercontent.com/29718932/219623668-053ed52f-342f-4dc1-8669-a05f91b3c778.png">
Tooltip:
<img width="1157" alt="image" src="https://user-images.githubusercontent.com/29718932/219623546-819693f2-4e81-4690-bc97-4be397b9bf9a.png">
